prov/efa/src/rxr/rxr_cq.c

```diff
@@ -786,38 +795,46 @@ void rxr_cq_write_tx_completion(struct rxr_ep *ep,
 	return;
 }
 
-void rxr_cq_handle_tx_completion(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry)
+/**
+ * @brief This functions handles completion process after all data bytes of a message are sent 
+ *  	  and the receiver/requestor has acknowledged that the sent bytes were received. 
+ *
+ * @param[in]	ep		end point
+ * @param[in]	op_entry	tx/rx entry that contains information of the data send operation  
+ */
+void rxr_cq_handle_send_completion(struct rxr_ep *ep, struct rxr_op_entry *op_entry)
 {
-	if (tx_entry->state == RXR_TX_SEND)
-		dlist_remove(&tx_entry->entry);
+	struct rxr_op_entry *rx_entry;
+
+	if (op_entry->state == RXR_TX_SEND)
+		dlist_remove(&op_entry->entry);
 
-	if (tx_entry->cq_entry.flags & FI_READ) {
+	if (op_entry->cq_entry.flags & FI_READ) {
 		/*
 		 * This is on responder side of an emulated read operation.
 		 * In this case, we do not write any completion.
-		 * The TX entry is allocated for emulated read, so no need to write tx completion.
+		 * The entry is allocated for emulated read, so no need to write tx completion.
 		 * EFA does not support FI_RMA_EVENT, so no need to write rx completion.
 		 */
-		struct rxr_rx_entry *rx_entry = NULL;
-
-		rx_entry = ofi_bufpool_get_ibuf(ep->rx_entry_pool, tx_entry->rma_loc_rx_id);
-		assert(rx_entry);
-		assert(rx_entry->state == RXR_RX_WAIT_READ_FINISH);
+		assert(op_entry->type == RXR_RX_ENTRY);
+		rx_entry = op_entry;
 		rxr_release_rx_entry(ep, rx_entry);
```

Comment:
I think you need to return here.

Reply:
Yeah the return is on line 822 in the changed version of the file.